pkg/disk/v1alpha2/disk.go

```diff
@@ -0,0 +1,251 @@
+func IsType(diskType string) Predicate {
+	return func(d *Disk) bool {
+		return d.Object.GetLabels()[string(apis.NdmDiskTypeCPK)] == diskType
+	}
+}
+
+// IsValidPoolTopology returns true if the topology is valid.
+func IsValidPoolTopology(poolType string, diskCount int) bool {
+	return DefaultDiskCount[poolType]%diskCount == 0
+}
+
+// IsDiskType returns true if the spc is of disk type.
+func IsDiskType() Predicate {
+	return func(d *Disk) bool {
+		return d.Object.GetLabels()[string(apis.NdmDiskTypeCPK)] == string(apis.TypeDiskCPV)
+	}
+}
+
+// GetNodeName returns the node name to which the disk is attached
+func (d *Disk) GetNodeName() string {
+	return d.Object.GetLabels()[string(apis.HostNameCPK)]
+}
+
+// Filter will filter the disk instances if all the predicates succeed against that spc.
+func (l *DiskList) Filter(p ...Predicate) *DiskList {
+	var plist predicateList
+	plist = append(plist, p...)
+	if len(plist) == 0 {
+		return l
+	}
+
+	filtered := NewListBuilder().List()
+	for _, spcAPI := range l.ObjectList.Items {
+		spcAPI := spcAPI // pin it
+		Disk := BuilderForAPIObject(&spcAPI).Disk
+		if plist.all(Disk) {
+			filtered.ObjectList.Items = append(filtered.ObjectList.Items, *Disk.Object)
+		}
+	}
+	return filtered
+}
+
+// NewBuilder returns an empty instance of the Builder object.
+func NewBuilder() *Builder {
+	return &Builder{
+		Disk: &Disk{&apis.Disk{}},
+	}
+}
+
+// BuilderForObject returns an instance of the Builder object based on spc object
+func BuilderForObject(Disk *Disk) *Builder {
+	return &Builder{
+		Disk: Disk,
+	}
+}
+
+// BuilderForAPIObject returns an instance of the Builder object based on spc api object.
+func BuilderForAPIObject(spc *apis.Disk) *Builder {
+	return &Builder{
+		Disk: &Disk{spc},
+	}
+}
+
+// GetDeviceID returns the device ID of the disk.
+func (d *Disk) GetDeviceID() string {
+	var deviceID string
+	if len(d.Object.Spec.DevLinks) != 0 && len(d.Object.Spec.DevLinks[0].Links) != 0 {
+		deviceID = d.Object.Spec.DevLinks[0].Links[0]
+	} else {
+		deviceID = d.Object.Spec.Path
+	}
+	return deviceID
+}
+
+// Build returns the Disk object built by this builder.
+func (sb *Builder) Build() *Disk {
+	return sb.Disk
+}
+
+// NewListBuilder returns a new instance of ListBuilder object.
+func NewListBuilder() *ListBuilder {
+	return &ListBuilder{DiskList: &DiskList{ObjectList: &apis.DiskList{}}}
+}
+
+// WithList builds the list based on the provided *DiskList instances.
+func (b *ListBuilder) WithList(pools *DiskList) *ListBuilder {
+	if pools == nil {
+		return b
+	}
+	b.DiskList.ObjectList.Items = append(b.DiskList.ObjectList.Items, pools.ObjectList.Items...)
+	return b
+}
+
+// WithAPIList builds the list based on the provided *apis.CStorPoolList.
+func (b *ListBuilder) WithAPIList(pools *apis.DiskList) *ListBuilder {
+	if pools == nil {
+		return b
+	}
+	for _, pool := range pools.Items {
+		pool := pool //pin it
+		b.DiskList.ObjectList.Items = append(b.DiskList.ObjectList.Items, pool)
```

Comment:
Will below work & avoid for loop completely?
```go
b.DiskList.ObjectList.Items = append(b.DiskList.ObjectList.Items, disks.Items...)
```

Reply:
yes. done